plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -666,17 +712,145 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
+-}
+
+-- | Do your best to try to extract a source span from a tick
+-- See Note [What source locations to cover]
+getSourceSpan :: Maybe GHC.ModBreaks -> GHC.GenTickish pass -> Maybe GHC.RealSrcSpan
+getSourceSpan _ GHC.SourceNote{GHC.sourceSpan=src} = Just src
+getSourceSpan _ GHC.ProfNote{GHC.profNoteCC=cc} =
+  case cc of
+    GHC.NormalCC _ _ _ (GHC.RealSrcSpan sp) -> Just sp
+    GHC.AllCafsCC _ (GHC.RealSrcSpan sp)    -> Just sp
+    _                                       -> Nothing
+getSourceSpan mmb GHC.HpcTick{GHC.tickId=tid} = do
+  mb <- mmb
+  let arr = GHC.modBreaks_locs mb
+      range = Array.bounds arr
+  GHC.RealSrcSpan sp <- if Array.inRange range tid  then Just $ arr Array.! tid else Nothing
+  return sp
+getSourceSpan _ _ = Nothing
+
+-- | Obviously this function computes a GHC.RealSrcSpan from a CovLoc
+toCovLoc :: GHC.RealSrcSpan -> CovLoc
+toCovLoc sp = CovLoc (GHC.unpackFS $ GHC.srcSpanFile sp)
+                     (GHC.srcSpanStartLine sp)
+                     (GHC.srcSpanEndLine sp)
+                     (GHC.srcSpanStartCol sp)
+                     (GHC.srcSpanEndCol sp)
+
+-- Here be dragons:
+-- See Note [Tracking coverage and lazyness]
+-- See Note [Coverage order]
+-- | Annotate a term for coverage
+coverageCompile :: CompilingDefault uni fun m
+                => GHC.CoreExpr -- ^ The original expression
+                -> GHC.Type -- ^ The type of the expression
+                -> GHC.RealSrcSpan -- ^ The source location of this expression
+                -> PIRTerm uni fun -- ^ The current term (this is what we add coverage tracking to)
+                -> CoverageType -- ^ The type of coverage to do next
+                -> m (PIRTerm uni fun)
+coverageCompile originalExpr exprType src compiledTerm covT =
+  case covT of
+    -- Add a location covereage annotation to tell us "we've executed this piece of code"
+    LocationCoverage -> do
+      ann <- addLocationToCoverageIndex (toCovLoc src)
+      ty <- compileTypeNorm exprType
+      mkLazyTrace ty (T.pack . show $ ann) compiledTerm
+
+    -- Add two boolean coverage annotations to tell us "this boolean has been True/False respectively"
+    BooleanCoverage -> do
+      -- Check if the thing we are compiling is a boolean
+      CompileContext {ccNameInfo=nameInfo} <- ask
+      let bool = GHC.getName <$> Map.lookup ''Bool nameInfo -- TODO: This should always return a `Just` and it may be worthwhile to crash the compiler with an error if it doesn't
+          tyHeadName = GHC.getName <$> GHC.tyConAppTyCon_maybe exprType
+          headSymName = GHC.getName <$> findHeadSymbol originalExpr
+          isTrueOrFalse = case headSymName of
+            Nothing -> False
+            -- We have to do it this convoluted way because GHC decides that
+            -- `nm` is a *variable* but `'True` and `'False` are data constructors and
+            -- they are not equal as names even if `originalExpr` is e.g. the expression `True`
```

Comment:
That sounds wrong. I'm pretty sure we identify variable references to data constructors in precisely this way, so I think maybe you're doing something wrong.

Reply:
Could you point me at the code where you use simple equality to check that a variable name is a reference to a constructor? Because that's what we did first and it didn't work. We later did extensive experiments and found that if you use `'Con` in the plugin and you use `Con` in the source code the name of the first will not be equal to the name of the second because of the reason in the comment.

Reply:
https://github.com/input-output-hk/plutus/blob/master/plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs#L559

Reply:
That's super odd. Rewriting it this way works. Apparently `getName` before and after `idDetails` does not give you the same thing.